File: solutions/javascript/min-heap/03-peek/minHeap.js
class MinHeap {
  constructor() {
    this._items = [];
  }

  insert(value) {
    this._items.push(value);
    this._bubbleUp(this._items.length - 1);
  }

  peekMin() {
    if (this._items.length === 0) {
      return undefined;
    }
    return this._items[0];
  }

  _bubbleUp(index) {
    let current = index;
    while (current > 0) {
      const parent = Math.floor((current - 1) / 2);
      if (this._items[parent] <= this._items[current]) {
        break;
      }
      const temp = this._items[parent];
      this._items[parent] = this._items[current];
      this._items[current] = temp;
      current = parent;
    }
  }
}

module.exports = MinHeap;

File: solutions/java/min-heap/step02_heapify_up/MinHeap.java
import java.util.ArrayList;
import java.util.List;

public class MinHeap {
    private final List<Integer> items;

    public MinHeap() {
        this.items = new ArrayList<>();
    }

    public void insert(int value) {
        items.add(value);
        bubbleUp(items.size() - 1);
    }

    private void bubbleUp(int index) {
        int current = index;
        while (current > 0) {
            int parent = (current - 1) / 2;
            if (items.get(parent) <= items.get(current)) {
                break;
            }
            swap(parent, current);
            current = parent;
        }
    }

    private void swap(int i, int j) {
        int temp = items.get(i);
        items.set(i, items.get(j));
        items.set(j, temp);
    }
}

